File: 13-Algorithms_sorting/insertion_sort.rb
# frozen_string_literal: true

def insertion_sort(arr)
  return arr if arr.length < 2

  (arr.length - 1).times.each do |i|
    j = i + 1

    while arr[j] < arr[j - 1]
      break unless j.positive?

      arr[j], arr[j - 1] = arr[j - 1], arr[j]
      j -= 1
    end
  end

  arr
end

numbers = [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0]
p insertion_sort(numbers)

File: 13-Algorithms_sorting/insertion_sort.js
const insertionSort = arr => {
  for (let i = 0; i < arr.length; i++) {
    let j = i + 1;

    while (arr[j] < arr[j - 1]) {
      [arr[j], arr[j - 1]] = [arr[j - 1], arr[j]];
      j--;
    }
  }

  return arr;
};

const numbers = [99, 44, 6, 2, 1, 5, 63, 87, 283, 4, 0];
console.log(insertionSort(numbers));
